### PR TITLE
Build wheel

### PR DESCRIPTION
**Pull request by Czaki**
I would like to add build wheel for psutil to allow easy install on non gcc macos and linux system. 

All wheels are build without problem, but when I add tests it fails. I found that part of test need to be disabled (because they are disabled on travis). 

I choose github workflows, because they allow to create artefacts (on travis there is need to upload to remote server) and not need to create additional account like on AzurePipelines.

`cibuildwheel` is really nice tool and reduce maintainty of  building wheel process. most often to upgrade version of cibuildwheel. 

I see discussion in #824 and think that I can help with fast wheel creation, but I do not know psutil enough weel to understnad if some test should be blacklisted if run on Macos 10.15 or inside Docker container.

### CODE REVIEW

**Comment by giampaolo** [comment]
Nice, this looks promising. On what Linux distro is this built? 
Without going back to re-read the whole #824 (I forgot details) I seem to remember we have 2 problems:
- 1 wheels must be built on CentOS 6 in order to be more compatible
- that means we'll have to change some stuff in `psutil/_psutil_linux.c` first
What Linux distro are these wheels built on?
Also, where are the artifacts stored and how do I get to them? (I suppose I'll have to download them locally, and the re-upload them on PYPI)

**Comment by Czaki** [comment]
Currently it is build on Centos 7 but I will switch it to Centos6 in next commit. 
These are `manylinux2010` image which is `Centos 7` based. 

On Centos 6 I register few more erors.

Wheels will be avaliable in actions. You can see here https://github.com/Czaki/psutil/actions 
After merge it will be avaliable in actions for this repository.

**Comment by Czaki** [comment]
pypy wheels will be build on centos7 because they not offer centos6 docker based images.

**Comment by giampaolo** [comment]
> On Centos 6 I register few more erors.

I suppose that's because of `_psutil_linux.c`: the parts relying of pre-processor checks (`#ifdef`s) will either have to be replaced with a counterpart detecting kernel / glibc version at runtime in C, or in Python by using ctypes. It's one of the blockers. For now I suggest to leave that alone (if it compiles) and ignore the test failures related to that.

**Comment by Czaki** [comment]
As far as I know I try to response on your answer to manylinux. 

There are three manylinux tags

* manylinux1
* manylinux2010 (most distro released after 2010 supported)
* manylinux2014 (most distro released after 2014 supported) 
 
and pip know host system glibc version and then check if there is available wheel with enough low version glibc. So if hos system has to low version of glibc then it will build from source. 

Main weekness of manylinux2010 platform tag is that it need pip in version 19.0.

The idea of building manylinux wheel is to build them inside docker container. manylinux team create docker images `manylinux1` - centos 6, `manylinux2010` - centos 7 , `manylinux2014` -  centos 8.  So there is no problem that travis support centos 7, because everything is build inside docker container. 

I'm contributor to `cibuildwheel` package and can try explain more technical details.

**Comment by giampaolo** [comment]
> and pip know host system glibc version and then check if there is available wheel with enough low version glibc. So if hos system has to low version of glibc then it will build from source.

Sweet. In that case I think we should stick with CentOS 7. For the record, I've been using a similar policy for Windows XP (supported from sources only, but now it's gone for good). 

Question, in case you know: what happens if the wheel does not support a recent distro? Is pip able to know that in advance or retry by using the tarball? My doubt is if by distrubuting wheels there's a chance that we break some installations which right now work fine because the tarball is the only thing available on PYPI.

**Comment by Czaki** [comment]
> Question, in case you know: what happens if the wheel does not support a recent distro? Is pip able to know that in advance or retry by using the tarball? My doubt is if by distributing wheels there's a chance we break some installations that now work fine because the tarball is the only thing currently available on PYPI.

What did you men by not support? wheels are checked with `auditwheel` that verify that they depend only on few libraries and should run on every more modern system. 

I do not meet case when wheel works only with old system and does not work with modern one. 

Win wheel name there is platform tag. `win` macos_version` or `manylinux*` pip know which is maximum platform tag that is supported.

**Comment by Czaki** [comment]
@giampaolo I I should manually mark some test as not use in build, or wait on signal from you that I should use more modern version?

**Comment by giampaolo** [comment]
Is it using CentOS 6 or 7?
From this failure:
https://github.com/Czaki/psutil/runs/671305932?check_suite_focus=true

```
======================================================================
FAIL: psutil.tests.test_contracts.TestAvailConstantsAPIs.test_linux_rlimit
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/project/psutil/tests/test_contracts.py", line 91, in test_linux_rlimit
    ae(hasattr(psutil.Process, "rlimit"), hasit)
AssertionError: False != True
```

I guess we're on 6?

**Comment by Czaki** [comment]
This build is on centos 7 https://github.com/Czaki/psutil/actions/runs/103672866
This is on centos 6 https://github.com/Czaki/psutil/actions/runs/103718944

This is two line switch. Go back to centos 7?

**Comment by Czaki** [comment]
What to do with error `OSError: AF_UNIX path too long` ?

**Review by giampaolo — CHANGES_REQUESTED** [review]
> This is two line switch. Go back to centos 7?

Yes. I pushed a couple of commits which should fix some failures so you should rebase.

File: setup.py [hunk 1]
```diff
@@ -67,7 +67,11 @@
...
 
 
-extras_require = {}
+extras_require = {"test": [
+    "ipaddress; python_version < '3.0'",
+     "mock; python_version < '3.0'",
+     "pypiwin32; sys.platform == 'win32'",
+     "wmi; sys.platform == 'win32'"]}
 if sys.version_info[:2] <= (3, 3):
     extras_require.update(dict(enum='enum34'))
```
[the remark this answers is not in the document]
Comment (Czaki) [reply setup.py]: fixed

File: psutil/tests/__init__.py [hunk 2]
```diff
@@ -123,7 +123,8 @@
 TRAVIS = bool(os.environ.get('TRAVIS'))
 APPVEYOR = bool(os.environ.get('APPVEYOR'))
 CIRRUS = bool(os.environ.get('CIRRUS'))
-CI_TESTING = TRAVIS or APPVEYOR or CIRRUS
+GITHUB = bool(os.environ.get('CI', False))
```
[the remark this answers is not in the document]
Comment (Czaki) [reply psutil/tests/__init__.py]: ok

**Review by Czaki — COMMENTED** [review]
Still few errors (2 on mac 4 on linux) and unknown reason of windows fail

**Comment by giampaolo** [comment]
According to this failure:
```
FAIL: psutil.tests.test_contracts.TestAvailProcessAPIs.test_rlimit
----------------------------------------------------------------------
Traceback (most recent call last):
File "/project/psutil/tests/test_contracts.py", line 153, in test_rlimit
self.assertEqual(hasattr(psutil.Process, "rlimit"), LINUX)
AssertionError: False != True
```
...`Process.prlimit()` is not available, and that's because of this check: https://github.com/giampaolo/psutil/blob/2c113c6746d3558e47e575b4169ac15b636ddf57/psutil/_psutil_linux.c#L45-L48 

...which should be "true" on CentOS 7. Are you sure we're compiling against CentOS 7? The resulting wheels have `manylinux2010` in their name. Doesn't that mean CentOS 6?

**Comment by Czaki** [comment]
> ...which should be "true" on CentOS 7. Are you sure we're compiling against CentOS 7? The resulting wheels have `manylinux2010` in their name. Doesn't that mean CentOS 6?

argh... 

I do wrong counting `manylinux1` is from centos 5 so `manylinux2010` is form centos 6. so for centos 7 there is need to use manylinux2014. 

I add step with manylinux2014 build and it still fail.

**Comment by giampaolo** [comment]
OK, good. There are still some failures but the important one was to make `Process.rlimit()` available and that is fixed. manylinux2010 are still produced though. We should not run anything at all on CentOS6. Only run tests on CentOS7 and produce wheels on CentOS7, so basically I think we should only keep this section:
https://github.com/giampaolo/psutil/pull/1758/commits/39b5a0b2fff3d52f2d545c5bee1f73486a61d3de
...and remove the rest.

**Comment by Czaki** [comment]
Done. 
If I should rebase on current master?

It meanas that there will be no wheel for pypy for linux (they provide only manylinux 2010 docker image)

**Comment by giampaolo** [comment]
> It means that there will be no wheel for pypy for linux (they provide only manylinux 2010 docker image)

Mmmm. OK, I just downloaded the zip and indeed I see some wheels produced for PYPY with 2010/CentOS-6:

* psutil-5.7.1-pp36-pypy36_pp73-win32.whl
* psutil-5.7.1-pp36-pypy36_pp73-manylinux2010_x86_64.whl
* psutil-5.7.1-pp36-pypy36_pp73-macosx_10_9_x86_64.whl
* psutil-5.7.1-pp27-pypy_73-manylinux2010_x86_64.whl
* psutil-5.7.1-pp27-pypy_73-macosx_10_9_x86_64.whl

I have mixed feelings about PYPY. I consider it low priority, but since this basically comes for free maybe we can decide to either 1) accept that `Process.prlimit()` won't be available on PYPY + Linux 2) provide an additional pure-Python implementation of `prlimit()` which uses `ctypes`. I will think about that later. Let's not worry about that for now.  

Also I have a couple of questions about OSX:
* a lot of python versions are missing on OSX. I think it's because this script depends on `.travis.yml` right? In that case I think we should extend `.travis.yml` to also include python 2.7, 3.5, 3.6, 3.7, 3.8 on OSX.
* It looks like the wheels on OSX are produced on macosx 10.9. Will they work on more recent OSX versions?

As for this failure:

```
2020-05-16T10:18:31.7778978Z psutil.tests.test_connections.TestSystemWideConnections.test_multi_sockets_procs ... Traceback (most recent call last):
2020-05-16T10:18:31.7780948Z   File "/tmp/@psutil-100-rgd3uhff", line 2, in <module>
2020-05-16T10:18:31.7781195Z     from psutil.tests import create_sockets
2020-05-16T10:18:31.7781707Z ImportError: No module named 'psutil'
2020-05-16T10:18:31.7797725Z Traceback (most recent call last):
2020-05-16T10:18:31.7798318Z   File "/tmp/@psutil-100-s2bbb_6w", line 2, in <module>
2020-05-16T10:18:31.7798447Z     from psutil.tests import create_sockets
```
I think you can try this change:

```diff
diff --git a/psutil/tests/__init__.py b/psutil/tests/__init__.py
index 9667b5d5..33d5d2fd 100644
--- a/psutil/tests/__init__.py
+++ b/psutil/tests/__init__.py
@@ -199,7 +199,9 @@ def _get_py_exe():
         else:
             return exe
 
-    if MACOS:
+    if GITHUB_WHEELS:
+        return 'python'
+    elif MACOS:
         exe = \
             attempt(sys.executable) or \
             attempt(os.path.realpath(sys.executable)) or \
~/svn/psutil {master}$ 
```

If it doesn't work don't worry and I'll investigate it once we merge this. The few failures we have are issues with tests themselves and not psutil, so they are not important.

**Comment by Czaki** [comment]
> Also I have a couple of questions about OSX:
> 
>    * a lot of python versions are missing on OSX. I think it's because this script depends on `.travis.yml` right? In that case I think we should extend `.travis.yml` to also include python 2.7, 3.5, 3.6, 3.7, 3.8 on OSX.

I just download and see this files:
```
psutil-5.7.1-cp27-cp27m-macosx_10_9_x86_64.whl
psutil-5.7.1-cp35-cp35m-macosx_10_9_x86_64.whl
psutil-5.7.1-cp36-cp36m-macosx_10_9_x86_64.whl
psutil-5.7.1-cp37-cp37m-macosx_10_9_x86_64.whl
psutil-5.7.1-cp38-cp38-macosx_10_9_x86_64.whl
psutil-5.7.1-pp27-pypy_73-macosx_10_9_x86_64.whl
psutil-5.7.1-pp36-pypy36_pp73-macosx_10_9_x86_64.whl
```
which version is missed? 

>    * It looks like the wheels on OSX are produced on macosx 10.9. Will they work on more recent OSX versions?

macos 10.9 means that minimal version of system is macosx 10.9. Wheel should run on all newer version (from 10.15 only x86_64, but this wheel is already 64-bits)

**Comment by giampaolo** [comment]
Maybe this:
```diff
+    if GITHUB_WHEELS:
+        return os.path.realpath('python')
```

**Comment by Czaki** [comment]
`cibuildwheel` guarantee that python executable will be in path under `python` name. 

One problem which I see that if any test fail then wheel building fail. Because of this I add second group which build wheels without test. 

So I prefer to mark this test with x-fail to run see if there are no problem with other versions.

**Comment by giampaolo** [comment]
Yeah, I agree it is best to ignore test failures. I am already busy enough taking a look at "normal" (non-wheel) runs. =)
Anyway, this is great work, thanks. I didn't know github was capable of this. I'm going to merge this PR and try to adjust those minor failures on my own. I'll also have to see how to write a script to download the zip (by using github APIs I guess) for when I'll have to make a new release.

**Comment by giampaolo** [comment]
Python 2.7 wheels for Linux are missing due to this directive:
```
 CIBW_SKIP: "pp27-*win* cp27-*manylinux* pp-*manylinux*"
```
When I remove it I get this error:
```
    + echo 'cibuildwheel: python available on PATH doesn'\''t match our installed instance. If you have modified PATH, ensure that you don'\''t overwrite cibuildwheel'\''s entry or insert python above it.'
```
The issue disappears if I use CentOS 6 / manylinux2010. I've been struggling with this the whole day. :(

**Comment by Czaki** [comment]
Yes. manylinux2014 does not provide python 2.7 executable. 

You supprise me with this merge I have add few commit to branch but I plan to create next PR when I got working version. 
There. Pypy on macos add specific variable to environment for example.

**Comment by giampaolo** [comment]
Oh sorry, I thought you were done. And it was probably too soon anyway (the Windows error is very obscure). I have a PR I am going to merge soon which fixes some of the tests which were/are failing here.

> Yes. manylinux2014 does not provide python 2.7 executable.

Then we're stuck with 2010. That'll probably mean ctypes for `prlimit()` (not fun).

**Comment by giampaolo** [comment]
Merged #1761.

**Pull request merged**